### PR TITLE
feat: add balance checks for withdrawal

### PR DESCRIPTION
**Pull request by adu-web3**
## Description

This PR adds balance checks in two places:

1. in `LSTRestakingController.claimPrincipalFromImuachain` function before sending out the messages, so that it could fail early before sending messages that would definitely fail
2. in `UTXOGateway.withdrawPrincipal` and `UTXOGateway.withdrawReward` functions by checking against the dust threshold for BTC and DOGE

it also adds or changes some constant values:

1. `DESTINATION_GAS_LIMIT` is increased to 700_000 from original 500_000 to avoid the out-of-gas error for LZ executor
2. `BTC_DUST_THRESHOLD` and `DOGE_DUST_THRESHOLD` are added so that we could check against them
```solidity
/* ---------------------- Dust threshold for withdrawal --------------------- */
    // The minimum UTXO amount for bitcoin network is 546 satoshis, while user needs to pay for the withdrawal network
    // fee, which is likely to be 421 satoshis in average. So 546 * 2 = 1092 satoshis should be safe enough for the
    // withdrawal.
    uint256 public constant BTC_DUST_THRESHOLD = 1092;
    // The minimum UTXO amount for dogecoin network is 0.01 DOGE, while user needs to pay for the withdrawal network
    // fee, which might be 0.13 or even higher in average according to https://coin.space/dogecoin-network-fee/
    // so 0.5 DOGE = 50000000 should be safe enough for the withdrawal.
    uint256 public constant DOGE_DUST_THRESHOLD = 50_000_000;
 ```
 

## Summary by CodeRabbit

* **New Features**
  - Added safeguards preventing principal claims that would exceed total deposits.
  - Implemented minimum withdrawal dust thresholds to block uneconomical BTC and DOGE withdrawals.

* **Documentation**
  - Clarified destination gas-limit assumption in cross-chain gateway docs.

* **Tests**
  - Expanded test suite covering deposits, unlock/claim limits, cumulative unlock edge cases, and BTC/DOGE dust-threshold withdrawal behaviors.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: src/core/LSTRestakingController.sol [hunk 1]
```diff
@@ -52,7 +53,13 @@ abstract contract LSTRestakingController is
     {
         // If we can get the vault, the token cannot be VIRTUAL_STAKED_ETH_ADDRESS, so that staker cannot bypass the
         // beacon chain merkle proof check to withdraw natively staked ETH
-        _getVault(token);
+        IVault vault = _getVault(token);
+
+        // If the amount to be unlocked would exceed the total deposited amount for the user, we should fail early
+        // before sending the message to Imuachain.
+        if (vault.wouldUnlockPrincipalExceedDeposit(msg.sender, principalAmount)) {
+            revert Errors.VaultPrincipalExceedsTotalDeposit();
```
Comment (bwhour) [inline src/core/LSTRestakingController.sol]: why not unify the error  **_VaultPrincipalExceedsTotalDeposit_**  with  **_VaultTotalUnlockPrincipalExceedsDeposit_** ？

File: src/storage/UTXOGatewayStorage.sol [hunk 2]
```diff
@@ -148,6 +148,16 @@ contract UTXOGatewayStorage {
...
 
+    /* ---------------------- Dust threshold for withdrawal --------------------- */
+    // The minimum UTXO amount for bitcoin network is 546 satoshis, while user needs to pay for the withdrawal network
+    // fee, which is likely to be 421 satoshis in average. So 546 * 2 = 1092 satoshis should be safe enough for the
+    // withdrawal.
+    uint256 public constant BTC_DUST_THRESHOLD = 1092;
+    // The minimum UTXO amount for dogecoin network is 0.01 DOGE, while user needs to pay for the withdrawal network
+    // fee, which might be 0.13 or even higher in average according to https://coin.space/dogecoin-network-fee/
+    // so 0.5 DOGE = 50000000 should be safe enough for the withdrawal.
+    uint256 public constant DOGE_DUST_THRESHOLD = 50_000_000;
```
[the remark this answers is not in the document]
Comment (adu-web3) [reply src/storage/UTXOGatewayStorage.sol:182]: this is just a soft threshold for the withdrawal amount, to prevent users from specifying too little amount. And we would add minimum withdrawal amount check in frontend pages to adapt to the realtime network fee, so these too layers of safeguards should prevent user from losing money owing to specifying too little withdrawal amount. Besides, this soft threshold is unlikely to cause security issues for the imuachain restaking business, so I suppose using a constant and conservative soft threshold should be acceptable, though I can add some comment to elaborate on the extra frontend safeguards. Does this make sense @coderabbitai ?

[the remark this answers is not in the document]
Comment (adu-web3) [reply src/storage/UTXOGatewayStorage.sol:182]: @coderabbitai please help me draft the comment according to our conversations, to elaborate on the additional frontend safeguard and that this would only cause user losing small amount of money in worst cases but would not cause protocol security issues

File: src/core/LSTRestakingController.sol (hunk 1 above)
Comment (adu-web3) [reply src/core/LSTRestakingController.sol]: you are right, we should use a unified error

**Review by bwhour — APPROVED** [review]
the whole looks good.

File: src/storage/UTXOGatewayStorage.sol [hunk 3]
```diff
@@ -160,6 +160,33 @@ contract UTXOGatewayStorage {
...
+    //
+    // Two-layer validation approach:
+    // 1. Contract layer (this threshold): Static, conservative constants that act as a safety net
+    // 2. Frontend layer: Dynamic checks based on real-time network fees for optimal UX
+    //
+    // BTC_DUST_THRESHOLD rationale:
+    // The minimum UTXO amount for Bitcoin network is 546 satoshis. Accounting for typical withdrawal
+    // network fees (historically ~421 satoshis, though subject to market volatility), we use 546 * 2 =
+    // 1092 satoshis as a conservative threshold. The frontend provides additional dynamic validation
+    // based on current network conditions.
```
Comment (MaxMustermann2) [inline src/storage/UTXOGatewayStorage.sol:175]: Smart.

**Review by MaxMustermann2 — APPROVED** [review]

**Review by cloud8little — APPROVED** [review]
